icu4c/source/common/unicode/brkiter.h

```diff
@@ -93,6 +95,34 @@ U_NAMESPACE_BEGIN
  * are available at http://www.unicode.org/reports/tr14/ and
  * http://www.unicode.org/reports/tr29/.
  * <p>
+ * A boundary logically occurs between two adjacent code points. The boundary
+ * analysis functions in this file return the index of the code unit directly
+ * after a boundary - so a value of 0 means "the beginning of the string" and a
+ * value of u_strlen(string) means "the end of the string".
+ * <p>
+ * A span of text between two adjacent boundaries is called a "segment."
+ * The boundary analysis functions in this file behave correctly regardless of
+ * whether the current position is set to a boundary location or into the
+ * middle of a segment. When the current position points to a boundary, the two
+ * segments that straddle that boundary are called the "previous segment" and
+ * the "current segment."
+ * <p>
+ * If the current position is set to the middle of a surrogate pair, the
+ * boundary analysis functions in this file will move the current position to
+ * the beginning of the surrogate pair before any other processing. If the
+ * current position then points into the middle of a segment, these routines
+ * will move to the beginning or end of that segment. Otherwise, the current
+ * position points to a segment boundary, and these routines will move to the
+ * beginning of the previous segment or the end of the current segment.
+ * <p>
+ * The boundary analysis functions in this file may scan an unbounded number of
+ * characters around the current position or even the returned position. (One
+ * example is a long string of flag emoji. In order to know whether the current
```

Comment:
Because the current position is known to be a boundary `next()` can work efficiently with flag emojis - just move two forward. Backwards iteration, or positioning with `preceding()` or `following()` do run into the behavior described, though.

Combining mark sequences also can cause arbitrarily long scanning.